### PR TITLE
DPMSolverMultistep add `rescale_betas_zero_snr`

### PR DESCRIPTION
**Pull request by Beinsezii**
Effectively identical to the implementation for EulerDiscrete in https://github.com/huggingface/diffusers/pull/6024

TL;DR:
 - `rescale_zero_terminal_snr` function copied from DDIM
 - Fudge the last cumprod to `2**-24` from `0` when using zsnr so it doesn't produce an `inf` sigma
 - Upcast samples to `torch.float32` for the duration of `step()` because the zsnr sigmas are substantially less numerically stable and the performance hit is negligible.
   - Also has the side effect of producing *slightly* better images overall
 - `use_karras_sigmas=True` still produces strange results as it does in EulerDiscrete. Might be worth investigating separately?
 - Simple true/false UT for the new config value

Demo images
dpm_compare_labels

`use_karras_sigmas` also seems to just generally produce bad/noisy results even without ZSNR so I might not have the config set up correctly?

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_dpmsolver_multistep.py [hunk 1]
```diff
@@ -880,6 +930,11 @@ def step(
         if self.step_index is None:
             self._init_step_index(timestep)
 
+        # store old dtype because model_output isn't always the same it seems
+        return_dtype = sample.dtype
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_dpmsolver_multistep.py:934]: do you mean `model_output.dtype` isn't always the same as `sample.dtype` before the upcast?

**Review by yiyixuxu — COMMENTED** [review]
oh thanks
left a question about dtype 
also would be lovely if we can see some sample comparison:)

**Comment by yiyixuxu** [comment]
> use_karras_sigmas=True still produces strange results as it does in EulerDiscrete. Might be worth investigating separately?

yes separate issue please:)

File: src/diffusers/schedulers/scheduling_dpmsolver_multistep.py (hunk 1 above)
Comment (Beinsezii) [reply src/diffusers/schedulers/scheduling_dpmsolver_multistep.py]: Turns out that was because `model_output = self.convert_model_output(model_output, sample=sample)` ends up casting it to the sample's dtype. I moved the sample upcast after this call so `return_type` is no longer needed. Outputs are the same.

**Comment by yiyixuxu** [comment]
can you fix the quality test?

**Comment by Beinsezii** [comment]
Alright I added a single newline.

**Comment by yiyixuxu** [comment]
thank you!

**Comment by lachlan-nicholson** [comment]
> > use_karras_sigmas=True still produces strange results as it does in EulerDiscrete. Might be worth investigating separately?
> 
> yes separate issue please:)

Have we created a separate issue for ZeroSNR + Karras?
At a glance the max lambda is very large and could be clipped similar to #3314, potentially via `config.sigma_max`. 
This gives viable results but introduces a new parameter.
I'm left wondering what it means to apply both ZeroSNR and Karras modifiers at the same time.
